### ISSUE TITLE
Cannot disable parallelization

### ISSUE THREAD

**Issue by nhejazi** [wrathematics/coop#15]
Thanks for this package @wrathematics! It's been really useful to us over at https://github.com/PhilBoileau/cvCovEst, where we're testing a number of different covariance matrix estimators. As part of our implementations of these different estimators, we rely on `coop::covar` as an alternative to `stats::cov`. When testing in simulation experiments recently, we ran into issues that appear to suggest that `coop::covar` is performing some sort of C-level parallelization, in a way that I'm unable to disable (even with two distinct solutions appearing in other issues here), including when setting either of
```r
library(openblasctl)
openblas_set_num_threads(1)
```

```r
Sys.setenv(OMP_NUM_THREADS = "1")
```

As we note at https://github.com/PhilBoileau/cvCovEst/issues/18, this seems to only happen on Linux machines (including our HPC resources), though not on macOS. The behavior appears preserved across recent R versions (i.e., both 3.6.X and 4.0.2). Is there something we're missing or perhaps some intended behavior of `coop::covar`?

**Comment by wrathematics** [comment]
Thanks for the kind words.

Unfortunately you can't set `OMP_NUM_THREADS` with `Sys.setenv()`. You can launch R via 

```
OMP_NUM_THREADS=1 R
```

and that will work. It is possible to control the number of threads from an active R session, but not with `Sys.setenv()`. Looks like this [CRAN package](https://cran.r-project.org/web/packages/OpenMPController/index.html) might be of interest. Setting `OMP_NUM_THREADS` will manage both the package's OpenMP usage as well as OpenBLAS/MKL/etc.

I'm assuming the reason it's not happening on macs is because the package is not built with OpenMP support in that case, but I'm not sure.

**Comment by nhejazi** [comment]
Thanks for the quick reply. I recalled the use of `Sys.setenv()` for `OMP_NUM_THREADS` in a similar (but completely unrelated) issue related to `xgboost`'s parallelization scheme. It seemed to work there, but I guess the how the environment variable affects multithreading differs in these two situations.

I had seen a comment you'd made about `OMP_NUM_THREADS=1 R` but I originally avoided this since it would have meant that users of our package might've been required to start the R repl in a way that might be unfamiliar to them, but this is probably a minor point not worth considering immediately. To clarify, would you expect
```bash
export OMP_NUM_THREADS='1'  # https://github.com/wrathematics/coop/issues/13
export MKL_THREADING_LAYER='sequential'  # https://github.com/HenrikBengtsson/future/issues/405

R CMD BATCH --no-save --no-restore \
  ...
```
to work? I'll try out `OMP_NUM_THREADS=1 R` locally but, more importantly, I'd like to get our simulations running in parallel in a batch submission on our HPC cluster. I believe the above should be equivalent but there might be a subtlety I'm missing.

Thanks also for pointing out that CRAN package, looks like that might be the right solution here since we seemed to separately run into issues with `future`-style parallelization, possibly due to MKL multithreading. Thanks for any help.

**Comment by wrathematics** [comment]
Sure, that should work if you're using bash or something similar for your shell.

I think all the BLAS libraries will respect `OMP_NUM_THREADS` so you shouldn't need the second line. Usually they have their own goofy way of disabling parallelism without disabling all OMP threading, but I think you want it all disabled. In that case setting the number of OMP threads to 1 should suffice.

**Comment by nhejazi** [comment]
Thanks for confirming, it's helped us out a lot. From trying `OpenMPController::omp_set_num_threads(1)`, I found that this fixed issues I was seeing locally (in combination with your `openblastctl::openblas_set_num_threads(1)`, though not sure if this is redundant). Setting the env var in the batch script didn't seem to work on our HPC cluster, but I'll try it again to confirm. I had no idea that `Sys.setenv(OMP_NUM_THREADS='1')` wasn't doing the job, so this was helpful again.

Anyway, seems like the issue doesn't have to do with `coop::covar` though that's where we noticed it in our debugging attempts. I'll close this; thanks again for the help.
